src/coreComponents/denseLinearAlgebra/denseLASolvers.hpp

```diff
@@ -0,0 +1,349 @@
+ * @tparam SOL_TYPE The type of the solution vector `x`.
+ * @param[in,out] A The matrix to be transformed into an upper triangular matrix. Modified in place.
+ * @param[in,out] b The right-hand side vector. Modified in place to reflect the transformed system.
+ * @param[out] x The solution vector. The result of solving the system `Ax = b`.
+ */
+template< std::ptrdiff_t N, 
+          typename MATRIX_TYPE,
+          typename RHS_TYPE,
+          typename SOL_TYPE >
+GEOS_HOST_DEVICE 
+inline
+void solveGaussianElimination( MATRIX_TYPE & A, RHS_TYPE & b, SOL_TYPE && x )
+{
+  static_assert( N > 0, "N must be greater than 0." );
+  internal::checkSizes< N, N >( matrix );
+  internal::checkSizes< N >( b ); 
+  internal::checkSizes< N >( x );
+  
+
+  // Step 1: Transform  into an upper triangular matrix 
+  
+  // 1.a. Find the pivot
+  for (std::ptrdiff_t i = 0; i < N; ++i)
+  {
+    std::ptrdiff_t max_row = i;
+    for (std::ptrdiff_t k = i + 1; k < N; ++k)
+    {
+      if (std::abs(A[k][i]) > std::abs(A[max_row][i]))
+      {
+        max_row = k;
+      }
+    }
+
+    // 1.b. Swap rows
+    for (std::ptrdiff_t k = i; k < N; ++k)
+    {
+      std::swap(A[i][k], A[max_row][k]);
+    }
+    std::swap(b[i], b[max_row]);
+
+    GEOS_ERROR_IF_LT_MSG( LvArray::math::abs(A[i][i]), LvArray::NumericLimits< real64 >::epsilon, "Singular matrix." );
+
+    // 1.c Eliminate entries below the pivot
+    for (std::ptrdiff_t k = i + 1; k < N; ++k)
+    {
+      real64 const scaling = A[k][i] / A[i][i];
+      for (std::ptrdiff_t j = i; j < N; ++j)
+      {
+        A[k][j] -= scaling * A[i][j];
+      }
+      b[k] -= scaling * b[i];
+    }
+  }
+
+  // Step 2: Backward substitution
+  solveUpperTriangularSystem<N>( A, b, std::forward<N>(x) )
+}
+
+/**
+ * @brief Const version of the solveGaussianElimination function.
+ * 
+ * This function solves a linear system using Gaussian elimination, without modifying the original matrix `A` 
+ * and vector `b`. It creates copies of `A` and `b`, performs the elimination, and then solves the system.
+ * 
+ * @tparam N The size of the square matrix `A`.
+ * @tparam MATRIX_TYPE The type of the matrix `A`.
+ * @tparam RHS_TYPE The type of the right-hand side vector `b`.
+ * @tparam SOL_TYPE The type of the solution vector `x`.
+ * @param[in] A The constant matrix representing the coefficients of the system.
+ * @param[in] b The constant right-hand side vector.
+ * @param[out] x The solution vector. The result of solving the system `Ax = b`.
+ */
+template< std::ptrdiff_t N, 
+          typename MATRIX_TYPE,
+          typename RHS_TYPE,
+          typename SOL_TYPE >
+GEOS_HOST_DEVICE 
+inline
+void solveGaussianElimination( MATRIX_TYPE const & A, RHS_TYPE const & b, SOL_TYPE && x )
+{
+  real64[N][N] A_copy{};
+  real64[N] b_copy{};
+
+  for(std::ptrdiff_t i=0; i < N; ++j)
+  {
+    b_copy[i] = b[i];
+    for( std::ptrdiff_t j=0; j < N; ++j )
+    {
+      A_copy[i][j] = A[i][j];
+    };
+  };
```

Comment:
Could we add a flag to optionally copy `A` and `b`, the default implementation overwriting the provided matrix and rhs?

Reply:
The option would be irrelevant for the `2x2` and `3x3` cases.

Reply:
Like instead of defining the `const` and `non-const` version of this?

Reply:
Precisely

Reply:
I find having a `const` and `non-const `version more elegant than a flag but we can have just the non-const version with a flag o another template parameter.

Reply:
Actually, doing what you suggested removed some code duplications.